### ISSUE TITLE
React.StrictMode broke this library (fixed)

### ISSUE THREAD

**Issue by HenrikBechmann** [HenrikBechmann/react-infinite-grid-scroller#14]
I'm working on a demo site, using RIGS outside the dev environment for the first time, and ... surprise!

it turns out that as of React 18, in dev mode, with React.StrictMode, react 'simulates' an unmount right after the first mount, and then continues running the hook function code as though nothing had happened.

Well, in the heavily async environment of an infinite scroller, it's important to tell whether a component is still mounted when changing state. So I set 

`const isMountedRef = useRef(true)`

and then set that to false with

~~~
useEffect(() => {
    return () => {isMountedRef.current = false}
}, [])
~~~

... just like a bazillion people advise.

But with the unmount simulation (trying to prepare us for their own asynchronous plans), then this unmount setting happens right at the start, without an actual unmount.

In short. It completely breaks the library. :-(

The workaround is to not use React.StrictMode. Then everything works.

I'm looking into this further, and frankly I'm pretty unhappy about this, since the return function of a useEffect without data dependencies was always touted as the hook component's version of onUnmountComponent.

Sigh.

If anyone has any ideas, I'm all ears.

**Comment by HenrikBechmann** [comment]
OK, so it turns out that in React.StrictMode in development, not only causes a 'faux unmount', but also actually runs useEffect with no dependencies *twice*. So my workaround:

~~~
const isMountedRef = useRef(true)

useEffect(() => {
    isMountedRef.current = true // to compensate for faux unmount in React.StrictMode
    return () => {isMountedRef.current = false}
}, [])

~~~

I hope and rather expect that happens fast enough to avoid any processes being incorrectly cancelled while `isMountedRef.current` is incorrectly set to `false`.

**Comment by HenrikBechmann** [comment]
There's an active discussion about all this going on here: https://github.com/facebook/react/issues/24670#issuecomment-1280222217
